src/native/containers/dn-vector.h

```diff
@@ -0,0 +1,463 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#ifndef __DN_VECTOR_H__
+#define __DN_VECTOR_H__
+
+#include "dn-utils.h"
+#include "dn-allocator.h"
+
+typedef int32_t (DN_CALLBACK_CALLTYPE *dn_vector_compare_func_t) (const void *a, const void *b);
+typedef bool (DN_CALLBACK_CALLTYPE *dn_vector_equal_func_t) (const void *a, const void *b);
+typedef void (DN_CALLBACK_CALLTYPE *dn_vector_for_each_func_t) (void *data, void *user_data);
+typedef void (DN_CALLBACK_CALLTYPE *dn_vector_dispose_func_t) (void *data);
+
+
+#define DN_DEFINE_VECTOR_T_STRUCT(name, type) \
+typedef struct _ ## name ## _t name ## _t; \
+struct _ ## name ## _t { \
+	type* data; \
+	uint32_t size; \
+	struct { \
+		uint32_t _element_size; \
+		uint32_t _capacity; \
+		uint32_t _attributes; \
+		dn_allocator_t *_allocator; \
+	}_internal; \
+}; \
+typedef struct _ ## name ## _it_t name ## _it_t; \
+struct _ ## name ## _it_t { \
+	uint32_t it; \
+	struct { \
+		name ## _t *_vector; \
+	} _internal; \
+}; \
+typedef struct _ ## name ## _result_t name ## _result_t; \
+struct _ ## name ## _result_t { \
+	bool result; \
+	name ## _it_t it; \
+}; \
+typedef struct _ ## name ## _custom_params_t name ## _custom_alloc_params_t; \
+typedef struct _ ## name ## _custom_params_t name ## _custom_init_params_t; \
+struct _ ## name ## _custom_params_t { \
+	dn_allocator_t *allocator; \
+	uint32_t element_size; \
+	uint32_t capacity; \
+	uint32_t attributes; \
+};
+
+DN_DEFINE_VECTOR_T_STRUCT(dn_vector, uint8_t);
+
+typedef enum {
+	DN_VECTOR_ATTRIBUTE_INIT_MEMORY = 0x1
+} dn_vector_attribute;
+
+static inline void
+dn_vector_it_advance (
+	dn_vector_it_t *it,
+	ptrdiff_t n)
+{
+	DN_ASSERT (it && ((it->it + n) < UINT32_MAX));
+	it->it = it->it + (int32_t)n;
+}
+
+static inline dn_vector_it_t
+dn_vector_it_next (dn_vector_it_t it)
+{
+	DN_ASSERT ((it.it + 1) < UINT32_MAX);
+	it.it++;
+	return it;
+}
+
+static inline dn_vector_it_t
+dn_vector_it_prev (dn_vector_it_t it)
+{
+	DN_ASSERT (it.it > 0);
+	it.it--;
+	return it;
+}
+
+static inline dn_vector_it_t
+dn_vector_it_next_n (
+	dn_vector_it_t it,
+	uint32_t n)
+{
+	DN_ASSERT ((it.it + n) < UINT32_MAX);
+	it.it += n;
+	return it;
+}
+
+static inline dn_vector_it_t
+dn_vector_it_prev_n (
+	dn_vector_it_t it,
+	uint32_t n)
+{
+	DN_ASSERT (it.it >= n);
+	it.it -= n;
+	return it;
+}
+
+static inline uint8_t *
+dn_vector_it_data (dn_vector_it_t it)
+{
+	DN_ASSERT (it._internal._vector && it._internal._vector->data);
+	return it._internal._vector->data + (it._internal._vector->_internal._element_size * it.it);
+}
+
+#define dn_vector_it_data_t(it, type) \
+	(type *)dn_vector_it_data ((it))
+
+static inline bool
+dn_vector_it_begin (dn_vector_it_t it)
+{
+	DN_ASSERT (it._internal._vector);
+	return it.it == 0;
+}
+
+static inline bool
+dn_vector_it_end (dn_vector_it_t it)
+{
+	DN_ASSERT (it._internal._vector);
+	return it.it == it._internal._vector->size;
+}
+
+#define DN_VECTOR_FOREACH_BEGIN(vector,var_type,var_name) do { \
```

Comment:
what do folks think of ordering the parameters the same way they would show up in a C#, Java, or javascript foreach statement? The collection is always the last parameter as in:
```C#
foreach(string s in list)
DN_VECTOR_FOREACH(var_type, var_name, vector)
```
I'm not sure if there is some other FOREACH precedent in C/C++. Pretty much all the code I always work with in C/C++ just uses an index into an array or pointer arithmetic but that may just be me using ancient conventions.

Reply:
Sounds reasonable to me, mainly put it as first parameter since that what the rest of the API's uses and resembles the "this" pointer, but for the iterate helper macros we can go either way.